File: android/src/main/kotlin/com/bluecortex/mobile/BlueCortexPowerSyncSupabasePlugin.kt
package com.bluecortex.mobile

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "BlueCortexPowerSyncSupabase")
class BlueCortexPowerSyncSupabasePlugin : Plugin() {

    private var manager: PowerSyncManager? = null
    private val watchCallbackIds = mutableMapOf<String, String>()
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    @PluginMethod
    fun initialize(call: PluginCall) {
        scope.launch {
            val config = call.getObject("config")
            if (config == null) {
                call.reject("Missing config parameter")
                return@launch
            }

            val powersyncUrl = config.getString("powersyncUrl")
            val supabaseUrl = config.getString("supabaseUrl")
            val supabaseAnonKey = config.getString("supabaseAnonKey")
            val schemaArray = config.getJSONArray("schema")

            if (powersyncUrl == null || supabaseUrl == null || supabaseAnonKey == null || schemaArray == null) {
                call.reject("Missing required configuration parameters (powersyncUrl, supabaseUrl, supabaseAnonKey, schema)")
                return@launch
            }

            val dbFilename = config.optString("dbFilename", "powersync.db")

            // Parse schema from JSON
            val schemaDefinition = mutableListOf<Map<String, Any>>()
            for (i in 0 until schemaArray.length()) {
                val tableObj = schemaArray.getJSONObject(i)
                val tableMap = mutableMapOf<String, Any>()
                tableMap["name"] = tableObj.getString("name")

                val columnsArray = tableObj.getJSONArray("columns")
                val columns = mutableListOf<Map<String, String>>()
                for (j in 0 until columnsArray.length()) {
                    val colObj = columnsArray.getJSONObject(j)
                    columns.add(mapOf(
                        "name" to colObj.getString("name"),
                        "type" to colObj.getString("type")
                    ))
                }
                tableMap["columns"] = columns

                if (tableObj.has("indexes")) {
                    val indexesArray = tableObj.getJSONArray("indexes")
                    val indexes = mutableListOf<Map<String, Any>>()
                    for (j in 0 until indexesArray.length()) {
                        val indexObj = indexesArray.getJSONObject(j)
                        val indexMap = mutableMapOf<String, Any>()
                        indexMap["name"] = indexObj.getString("name")

                        val indexColumnsArray = indexObj.getJSONArray("columns")
                        val indexColumns = mutableListOf<Map<String, Any>>()
                        for (k in 0 until indexColumnsArray.length()) {
                            val icObj = indexColumnsArray.getJSONObject(k)
                            val icMap = mutableMapOf<String, Any>("name" to icObj.getString("name"))
                            if (icObj.has("ascending")) {
                                icMap["ascending"] = icObj.getBoolean("ascending")
                            }
                            indexColumns.add(icMap)
                        }
                        indexMap["columns"] = indexColumns
                        indexes.add(indexMap)
                    }
                    tableMap["indexes"] = indexes
                }

                schemaDefinition.add(tableMap)
            }

            try {
                manager = PowerSyncManager(
                    context = context,
                    powersyncUrl = powersyncUrl,
                    supabaseUrl = supabaseUrl,
                    supabaseAnonKey = supabaseAnonKey,
                    schemaDefinition = schemaDefinition,
                    dbFilename = dbFilename
                )
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to initialize PowerSync: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun connect(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized. Call initialize() first.")
                return@launch
            }

            try {
                mgr.connect()
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to connect: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val clearDatabase = call.getBoolean("clearDatabase", false) ?: false

            try {
                mgr.disconnect(clearDatabase)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to disconnect: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun setToken(call: PluginCall) {
        val mgr = manager
        if (mgr == null) {
            call.reject("PowerSync not initialized")
            return
        }

        val token = call.getString("token")
        if (token == null) {
            call.reject("Missing token parameter")
            return
        }

        mgr.setToken(token)
        call.resolve()
    }

    @PluginMethod
    fun execute(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val sql = call.getString("sql")
            if (sql == null) {
                call.reject("Missing sql parameter")
                return@launch
            }

            val parameters = call.getArray("parameters")?.toList<Any>() ?: emptyList()

            try {
                val results = mgr.execute(sql, parameters)
                val ret = JSObject()
                ret.put("rows", JSArray(results))
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Query failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun getAll(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val sql = call.getString("sql")
            if (sql == null) {
                call.reject("Missing sql parameter")
                return@launch
            }

            val parameters = call.getArray("parameters")?.toList<Any>() ?: emptyList()

            try {
                val results = mgr.getAll(sql, parameters)
                val ret = JSObject()
                ret.put("rows", JSArray(results))
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Query failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun getOptional(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val sql = call.getString("sql")
            if (sql == null) {
                call.reject("Missing sql parameter")
                return@launch
            }

            val parameters = call.getArray("parameters")?.toList<Any>() ?: emptyList()

            try {
                val result = mgr.getOptional(sql, parameters)
                val ret = JSObject()
                if (result != null) {
                    ret.put("row", JSObject.fromJSONObject(org.json.JSONObject(result)))
                } else {
                    ret.put("row", JSObject.NULL)
                }
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Query failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun watch(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val sql = call.getString("sql")
            val callbackId = call.getString("callbackId")
            if (sql == null || callbackId == null) {
                call.reject("Missing sql or callbackId parameter")
                return@launch
            }

            val parameters = call.getArray("parameters")?.toList<Any>() ?: emptyList()

            try {
                val watchId = mgr.watch(sql, parameters) { results ->
                    val data = JSObject()
                    data.put("rows", JSArray(results))
                    notifyListeners("watch_$callbackId", data)
                }

                watchCallbackIds[watchId] = callbackId
                val ret = JSObject()
                ret.put("watchId", watchId)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("Watch failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun unwatch(call: PluginCall) {
        val mgr = manager
        if (mgr == null) {
            call.reject("PowerSync not initialized")
            return
        }

        val watchId = call.getString("watchId")
        if (watchId == null) {
            call.reject("Missing watchId parameter")
            return
        }

        mgr.unwatch(watchId)
        watchCallbackIds.remove(watchId)
        call.resolve()
    }

    @PluginMethod
    fun getSyncStatus(call: PluginCall) {
        val mgr = manager
        if (mgr == null) {
            call.reject("PowerSync not initialized")
            return
        }

        val status = mgr.getSyncStatus()
        call.resolve(JSObject.fromJSONObject(org.json.JSONObject(status)))
    }

    @PluginMethod
    fun getVersion(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val version = mgr.getVersion()
            val ret = JSObject()
            ret.put("version", version)
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun writeTransaction(call: PluginCall) {
        scope.launch {
            val mgr = manager
            if (mgr == null) {
                call.reject("PowerSync not initialized")
                return@launch
            }

            val sqlStatements = call.getArray("sql")?.toList<String>()
            if (sqlStatements == null) {
                call.reject("Missing sql array parameter")
                return@launch
            }

            val parameters = call.getArray("parameters")?.toList<List<Any>>() ?: emptyList()

            try {
                mgr.writeTransaction(sqlStatements, parameters)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Transaction failed: ${e.message}")
            }
        }
    }
}

File: ios/Sources/BlueCortexCapacitorPowersyncSupabasePlugin/BlueCortexPowerSyncSupabasePlugin.swift
import Foundation
import Capacitor

@objc(BlueCortexPowerSyncSupabasePlugin)
public class BlueCortexPowerSyncSupabasePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "BlueCortexPowerSyncSupabasePlugin"
    public let jsName = "BlueCortexPowerSyncSupabase"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "query", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getOptional", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "watch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unwatch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSyncStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "writeTransaction", returnType: CAPPluginReturnPromise)
    ]
    
    private var manager: PowerSyncManager?
    private var watchCallbackIds: [String: String] = [:]

    @objc public func initialize(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let config = call.getObject("config") else {
                call.reject("Missing config parameter")
                return
            }
            
            guard let powersyncUrl = config["powersyncUrl"] as? String,
                  let supabaseUrl = config["supabaseUrl"] as? String,
                  let supabaseAnonKey = config["supabaseAnonKey"] as? String,
                  let schema = config["schema"] as? [[String: Any]] else {
                call.reject("Missing required configuration parameters (powersyncUrl, supabaseUrl, supabaseAnonKey, schema)")
                return
            }
            
            let dbFilename = config["dbFilename"] as? String ?? "powersync.db"
            
            do {
                self.manager = try PowerSyncManager(
                    powersyncUrl: powersyncUrl,
                    supabaseUrl: supabaseUrl,
                    supabaseAnonKey: supabaseAnonKey,
                    schemaDefinition: schema,
                    dbFilename: dbFilename
                )
                
                call.resolve()
            } catch {
                call.reject("Failed to initialize PowerSync: \(error.localizedDescription)")
            }
        }
    }

    @objc public func connect(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized. Call initialize() first.")
                return
            }
            
            do {
                try await manager.connect()
                call.resolve()
            } catch {
                call.reject("Failed to connect: \(error.localizedDescription)")
            }
        }
    }

    @objc public func disconnect(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            let clearDatabase = call.getBool("clearDatabase") ?? false
            
            do {
                try await manager.disconnect(clearDatabase: clearDatabase)
                call.resolve()
            } catch {
                call.reject("Failed to disconnect: \(error.localizedDescription)")
            }
        }
    }

    @objc public func setToken(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let token = call.getString("token") else {
                call.reject("Missing token parameter")
                return
            }
            
            manager.setToken(token)
            call.resolve()
        }
    }

    @objc public func execute(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sql = call.getString("sql") else {
                call.reject("Missing sql parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [Any] ?? []
            
            do {
                let results = try await manager.execute(sql: sql, parameters: parameters)
                call.resolve(["rows": results])
            } catch {
                call.reject("Query failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func getAll(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sql = call.getString("sql") else {
                call.reject("Missing sql parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [Any] ?? []
            
            do {
                let results = try await manager.getAll(sql: sql, parameters: parameters)
                call.resolve(["rows": results])
            } catch {
                call.reject("Query failed: \(error.localizedDescription)")
            }
        }
    }

    /// Typed query – same as getAll, returns { rows }. Use from JS as PowerSync.query<YourType>({ sql, parameters }).
    @objc public func query(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sql = call.getString("sql") else {
                call.reject("Missing sql parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [Any] ?? []
            
            do {
                let results = try await manager.getAll(sql: sql, parameters: parameters)
                call.resolve(["rows": results])
            } catch {
                call.reject("Query failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func getOptional(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sql = call.getString("sql") else {
                call.reject("Missing sql parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [Any] ?? []
            
            do {
                let result = try await manager.getOptional(sql: sql, parameters: parameters)
                if let result = result {
                    call.resolve(["row": result])
                } else {
                    call.resolve(["row": NSNull()])
                }
            } catch {
                call.reject("Query failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func watch(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sql = call.getString("sql"),
                  let callbackId = call.getString("callbackId") else {
                call.reject("Missing sql or callbackId parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [Any] ?? []
            
            do {
                let watchId = try await manager.watch(sql: sql, parameters: parameters) { results in
                    self.notifyListeners("watch_\(callbackId)", data: ["rows": results])
                }
                
                self.watchCallbackIds[watchId] = callbackId
                call.resolve(["watchId": watchId])
            } catch {
                call.reject("Watch failed: \(error.localizedDescription)")
            }
        }
    }

    @objc public func unwatch(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let watchId = call.getString("watchId") else {
                call.reject("Missing watchId parameter")
                return
            }
            
            manager.unwatch(watchId: watchId)
            watchCallbackIds.removeValue(forKey: watchId)
            call.resolve()
        }
    }

    @objc public func getSyncStatus(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            let status = manager.getSyncStatus()
            call.resolve(status)
        }
    }

    @objc public func getVersion(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            let version = await manager.getVersion()
            call.resolve(["version": version])
        }
    }

    @objc public func writeTransaction(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let manager = self.manager else {
                call.reject("PowerSync not initialized")
                return
            }
            
            guard let sqlStatements = call.getArray("sql") as? [String] else {
                call.reject("Missing sql array parameter")
                return
            }
            
            let parameters = call.getArray("parameters") as? [[Any]] ?? []
            
            do {
                try await manager.writeTransaction(sqlStatements: sqlStatements, parameters: parameters)
                call.resolve()
            } catch {
                call.reject("Transaction failed: \(error.localizedDescription)")
            }
        }
    }
}
